cmd/rhelemeter-server/main.go

```diff
@@ -0,0 +1,441 @@
+		IdleConnTimeout:     30 * time.Second,
+	})
+
+	if o.Verbose {
+		transport = telemeter_http.NewDebugRoundTripper(o.Logger, transport)
+	}
+
+	forwardClient := &http.Client{
+		Transport: telemeter_http.NewInstrumentedRoundTripper("forward", transport),
+	}
+
+	if o.OIDCIssuer != "" {
+		provider, err := oidc.NewProvider(ctx, o.OIDCIssuer)
+		if err != nil {
+			return fmt.Errorf("OIDC provider initialization failed: %v", err)
+		}
+
+		ctx = context.WithValue(ctx, oauth2.HTTPClient,
+			&http.Client{
+				// Note, that e.g forward timeouts after 5s.
+				Timeout:   20 * time.Second,
+				Transport: telemeter_http.NewInstrumentedRoundTripper("oauth", transport),
+			},
+		)
+
+		cfg := clientcredentials.Config{
+			ClientID:     o.OIDCClientID,
+			ClientSecret: o.OIDCClientSecret,
+			TokenURL:     provider.Endpoint().TokenURL,
+		}
+
+		if o.OIDCAudienceEndpoint != "" {
+			cfg.EndpointParams = url.Values{"audience": []string{o.OIDCAudienceEndpoint}}
+		}
+
+		s := cfg.TokenSource(ctx)
+
+		forwardClient.Transport = &oauth2.Transport{
+			Base:   forwardClient.Transport,
+			Source: s,
+		}
+	}
+
+	var g run.Group
+	{
+		internal := http.NewServeMux()
+
+		// TODO: Refactor to not take *http.Mux
+		telemeter_http.DebugRoutes(internal)
+		telemeter_http.MetricRoutes(internal)
+		telemeter_http.HealthRoutes(internal)
+
+		r := chi.NewRouter()
+		r.Mount("/", internal)
+
+		r.Get("/", func(w http.ResponseWriter, req *http.Request) {
+			internalPathJSON, _ := json.MarshalIndent(Paths{Paths: []string{"/", "/metrics", "/debug/pprof", "/healthz", "/healthz/ready"}}, "", "  ")
+
+			w.Header().Add("Content-Type", "application/json")
+			if _, err := w.Write(internalPathJSON); err != nil {
+				level.Error(o.Logger).Log("msg", "could not write internal paths", "err", err)
+			}
+		})
+
+		s := &http.Server{
+			Handler: otelhttp.NewHandler(r, "internal", otelhttp.WithTracerProvider(tp)),
+		}
+
+		// Run the internal server.
+		g.Add(func() error {
+			if len(o.InternalTLSCertificatePath) > 0 {
+				if err := s.ServeTLS(internalListener, o.InternalTLSCertificatePath, o.InternalTLSKeyPath); err != nil && err != http.ErrServerClosed {
+					level.Error(o.Logger).Log("msg", "internal HTTPS server exited", "err", err)
+					return err
+				}
+			} else {
+				if err := s.Serve(internalListener); err != nil && err != http.ErrServerClosed {
+					level.Error(o.Logger).Log("msg", "internal HTTP server exited", "err", err)
+					return err
+				}
+			}
+			return nil
+		}, func(error) {
+			_ = s.Shutdown(context.TODO())
+			internalListener.Close()
+		})
+	}
+	{
+		external := chi.NewRouter()
+		external.Use(middleware.RequestID)
+
+		mux := http.NewServeMux()
+		telemeter_http.HealthRoutes(mux)
+		external.Mount("/", mux)
+
+		// v2 routes
+		{
+			v2ForwardClient := forwardClient
+
+			receiver, err := receive.NewHandler(o.Logger, o.ForwardURL, v2ForwardClient, prometheus.DefaultRegisterer, "tenantid", o.Whitelist, o.ElideLabels)
```

Comment:
tenantid should be taken from the flag so o.TenantID

Reply:
done